### PR TITLE
Inorder core fix

### PR DESCRIPTION
**Pull request by jj16791**
Within this PR, fixes to the use of the InOrder core have been made to allow for its compatibility with multi-cycle memory accesses and multiple execution units.

Two new pipeline units have been introduced, a blocking issue unit and an inorder stager. The first is similar to the dispatch/issue unit but forces inorder issuing and the latter is similar to the ROB but far more simplified and serves to track the order in which instructions are issued.

### CODE REVIEW

File: src/lib/pipeline/BlockingIssueUnit.cc [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+
+  for (size_t i = 0; i < registers.size(); i++) {
+    const auto& reg = registers[i];
+    // If the forwarded register is one depended on, supply value to uop at
+    // front of issue queue
+    if (dependent_ && reg == dependency_.first) {
+      auto& uop = issueQueue_.front();
+      uop->supplyOperand(dependency_.second, values[i]);
+      assert(uop->getSequenceId() == issueQueue_.front()->getSequenceId() &&
+             "Tried to early issue uop not at front of queue");
```
Comment (FinnWilkinson) [inline src/lib/pipeline/BlockingIssueUnit.cc:167]: Assert needs prefix

File: src/lib/pipeline/BlockingIssueUnit.cc [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+
+      dependent_ = false;
+      dependency_ = {};
+      break;
+    }
+  }
+}
+
+void BlockingIssueUnit::setRegisterReady(Register reg) {
+  scoreboard_[reg.type][reg.tag] = {true, -1};
```
Comment (FinnWilkinson) [inline src/lib/pipeline/BlockingIssueUnit.cc:225]: Again, using -1 in a uint64 could be problematic for very large simulation runs where this ID (i.e. uint64-max) is in use

**Review by FinnWilkinson — CHANGES_REQUESTED** [review]
Generally pretty clean, just some concerns over flushing using seqenceID mostly.

Some simple unit tests for the new pipeline units would be good too

File: src/include/simeng/models/inorder/Core.hh [hunk 3]
```diff
@@ -140,28 +145,49 @@ class Core : public simeng::Core {
...
   /** The writeback unit; writes uop results to the register files. */
   pipeline::WritebackUnit writebackUnit_;
 
+  /** The port allocator unit; allocates a port that an instruction will be
+   * issued from based on a defined algorithm. */
+  pipeline::PortAllocator& portAllocator_;
+
+  /** A queue of store address uops that have been retired that future store
+   * data uops can use to commit the load in the core's load/store queue unit.
+   */
```
[the remark this answers is not in the document]
Comment (jj16791) [reply src/include/simeng/models/inorder/Core.hh:188]: Reworded a bit more than suggested as it also referred to committing a load, not a store.

File: src/include/simeng/pipeline/BlockingIssueUnit.hh [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+namespace pipeline {
+
+/** An issue unit for an inorder pipelined processor. Reads instruction operands
+ * and performs scoreboarding. Issues instructions to the execution units once
+ * ready in program-order. */
+class BlockingIssueUnit {
+ public:
+  /** Construct an issue unit with references to input/output buffers,
+   * the register file, the port allocator, the and a description of the number
+   * of physical registers the scoreboard needs to reflect. */
```
[the remark this answers is not in the document]
Comment (jj16791) [reply src/include/simeng/pipeline/BlockingIssueUnit.hh:27]: removed "the"

File: src/include/simeng/pipeline/BlockingIssueUnit.hh [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+  /** The register availability scoreboard. An unsigned integer is used to set
+   * the instruction which made a register unavailable. */
+  std::vector<std::vector<std::pair<bool, uint64_t>>> scoreboard_;
+
+  /** A function to record the issue of an instruction to enable the tracking of
+   * in program-order instruction execution and writeback. */
+  std::function<void(const std::shared_ptr<Instruction>&)> recordIssue_;
+
+  /** A reference to the loadt/store queue. */
```
[the remark this answers is not in the document]
Comment (jj16791) [reply src/include/simeng/pipeline/BlockingIssueUnit.hh:95]: removed "t"

File: src/lib/models/inorder/Core.cc [hunk 6]
```diff
@@ -71,54 +117,99 @@ void Core::tick() {
...
+  bool euFlush = false;
+  uint64_t targetAddress = 0;
+  uint64_t lowestSeqId = 0;
+  for (const auto& eu : executionUnits_) {
+    if (eu.shouldFlush() && (!euFlush || eu.getFlushSeqId() < lowestSeqId)) {
+      euFlush = true;
+      lowestSeqId = eu.getFlushSeqId();
+      targetAddress = eu.getFlushAddress();
+    }
+  }
```
[the remark this answers is not in the document]
Comment (jj16791) [reply src/lib/models/inorder/Core.cc:173]: It should all be instructionID. The original stager unit worked off of sequenceID but I realised this issue and changed it to work with instruction objects akin to the ROB. Forgot to change all of this, good spot.

File: src/lib/models/inorder/Core.cc [hunk 7]
```diff
@@ -181,33 +280,67 @@ std::map<std::string, std::string> Core::getStats() const {
...
           {"context.switches", std::to_string(contextSwitches_)}};
 }
 
-void Core::raiseException(const std::shared_ptr<Instruction>& instruction) {
+void Core::raiseException(const std::shared_ptr<Instruction>& insn) {
+  // If an exception has already been generated by the pipeline, only replace
+  // the exceptionGeneratingInstruction_ if the passed instruction is younger
+  if (exceptionGenerated_) {
+    if (exceptionGeneratingInstruction_->getSequenceId() <
+        insn->getSequenceId()) {
```
[the remark this answers is not in the document]
Comment (jj16791) [reply src/lib/models/inorder/Core.cc:302]: The sequence ID worked better here as if two exceptions come from the same macro-op, you want to trigger the earlier one hence you need to work off of sequence ID not instruction ID.

File: src/include/simeng/pipeline/BlockingIssueUnit.hh [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+  const RegisterFileSet& registerFileSet_;
+
+  /** Whether the supply of a source operand is dependent on a currently
+   * executing instruction. */
+  bool dependent_ = false;
+
+  /** A pair representing a register dependency. The first entry in the pair is
+   * the register depended on, and the second is the operand index, within the
+   * dependent instruction, the dependency is associated with. */
+  std::pair<Register, uint16_t> dependency_;
```
[the remark this answers is not in the document]
Comment (jj16791) [reply src/include/simeng/pipeline/BlockingIssueUnit.hh:111]: It would register a new dependency. A more performant implementation would be to register all dependencies for the next instruction when it is considered for issue. I'll make that change.

File: src/lib/models/inorder/Core.cc [hunk 9]
```diff
@@ -9,31 +9,74 @@ namespace simeng {
...
+      for (YAML::Node gp : config["Execution-Units"][i]["Blocking-Groups"]) {
+        blockingGroups.push_back(gp.as<uint16_t>());
+      }
+    }
+    executionUnits_.emplace_back(
+        issuePorts_[i], completionSlots_[i],
+        [this](auto regs, auto values) {
+          issueUnit_.forwardOperands(regs, values);
+        },
+        [this](auto insn) { handleLoad(insn); },
```
[the remark this answers is not in the document]
Comment (jj16791) [reply src/lib/models/inorder/Core.cc:75]: Yes good spot, was left over from when `handleLoad` was more complex.

File: src/lib/models/inorder/Core.cc [hunk 10]
```diff
@@ -49,14 +92,17 @@ void Core::tick() {
...
         // Flush pipeline
         fetchUnit_.flushLoopBuffer();
         decodeUnit_.purgeFlushed();
-        executeUnit_.flush();
-        previousAddresses_ = std::queue<memory::MemoryAccessTarget>();
+        issueUnit_.flush();
+        staging_.flush();
+        for (auto& eu : executionUnits_) {
+          eu.flush();
+        }
```
[the remark this answers is not in the document]
Comment (jj16791) [reply src/lib/models/inorder/Core.cc:105]: Given that staging is empty, we know the LSQ will be too. It's like the OoO core with the ROB

[the remark this answers is not in the document]
Comment (jj16791) [reply src/lib/models/inorder/Core.cc:105]: I'll remove the EU/stager flush also as those are guaranteed to be emptied also

File: src/lib/ModelConfig.cc [hunk 11]
```diff
@@ -286,40 +286,46 @@ void ModelConfig::validate() {
...
-                  ExpectedValue::String)) {
-            uint16_t mappedGroup =
-                groupMapping_[euNode[subFields[1]][j].as<std::string>()];
-            blockingGroups.push(mappedGroup);
-            configFile_["Execution-Units"][i]["Blocking-Groups"][j] =
-                mappedGroup;
+        if (configFile_["Core"]["Simulation-Mode"].as<std::string>() ==
+            "inorderpipelined") {
+          invalid_ << "\t - Execution blocking groups are not compatible with "
+                      "an inorderpipelined Simulation Mode\n";
```
[the remark this answers is not in the document]
Comment (jj16791) [reply src/lib/ModelConfig.cc:292]: The flow of instructions in the EU must be inorder because of how we stall at the WB unit (staling the completion slots). With blocking groups, it's possible for uops to flow out of order within the EU and thus a uop gets stuck in the EU behind a younger-decoded uop waiting for that uop to enter the WB. 

A solution would be to instead queue all the instructions in the WB unit (creates issues with an ever-growing queue whilst the next uop to WB is executing which then has to be searched through to find the next program-ordered uop to WB) or to queue it within the EU unit (extra complexity all for the inorder core and has the same problem of search or ordering uops as the first solution). 

Additionally, unlike the OoO core, blocking groups within an in-order context seems weird. You'd be partially stalling entries in the EU for them to fully stalled at the pipeline at WB. You could also say that about the internal pipeline of the EU but it may help with pipeline bubbles when concerned with RAW/WAW dependencies.

File: src/lib/pipeline/DispatchIssueUnit.cc [hunk 12]
```diff
@@ -178,8 +178,6 @@ void DispatchIssueUnit::forwardOperands(const span<Register>& registers,
 
   for (size_t i = 0; i < registers.size(); i++) {
     const auto& reg = registers[i];
-    // Flag scoreboard as ready now result is available
-    scoreboard_[reg.type][reg.tag] = true;
```
[the remark this answers is not in the document]
Comment (jj16791) [reply src/lib/pipeline/DispatchIssueUnit.cc:182]: With the introduction of the additional complexity for the in-order core, it's not guaranteed that a uops values will be written to the reg file the cycle after they've been forwarded. Thus, we had to set the register to be ready (the logic of what's been commented on) at WB. Whilst this above logic works for the OoO in its current state and the aforementioned assumption holds, it may not exist in the future so I copied the logic for the OoO core also. It has the same emergent behaviour as the WB unit is ticked first of all the primary pipeline units.

TLDR; it was unsafe and worked on OoO model assumptions which may not hold in the future.

File: src/lib/pipeline/ExecuteUnit.cc [hunk 13]
```diff
@@ -215,6 +223,56 @@ void ExecuteUnit::purgeFlushed() {
...
+    }
+  }
+
+  if (replace && operationsStalled_.size() > 0) {
+    // Add uop to pipeline
+    auto& uop = operationsStalled_.front();
+    pipeline_.push_back({nullptr, tickCounter_ + uop->getLatency() - 1});
+    pipeline_.back().insn = std::move(uop);
+    operationsStalled_.front() = pipeline_.back().insn;
+  }
```
[the remark this answers is not in the document]
Comment (jj16791) [reply src/lib/pipeline/ExecuteUnit.cc:268]: When flushed, a now unblock operation should transition to the internal pipeline straight away. However, if we do it outside of the flush and with the tick, then it could be multiple cycles until it enters the pipeline as that logic is only triggered when the front of the pipeline is executed.

File: src/lib/ModelConfig.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (FinnWilkinson) [reply src/lib/ModelConfig.cc:292]: That all makes sense, thanks for the explanation

File: src/lib/pipeline/InOrderStager.cc [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+uint64_t InOrderStager::getNextSeqID() const {
+  // Return the next sequence ID that can undergo writeback logic, return -1 if
+  // none exist
+  if (issueOrderQueue_.size()) return issueOrderQueue_.front()->getSequenceId();
+  return -1;
+}
+
+void InOrderStager::recordRetired(uint64_t seqId) {
+  assert(issueOrderQueue_.front()->getSequenceId() == seqId &&
+         "Tried to record a retirement out of program order");
```
Comment (FinnWilkinson) [inline src/lib/pipeline/InOrderStager.cc:35]: Assert needs prefix

**Review by FinnWilkinson — CHANGES_REQUESTED** [review]
All looks good, just some very minor points

File: src/include/simeng/pipeline/BlockingIssueUnit.hh [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+  const RegisterFileSet& registerFileSet_;
+
+  /** Whether the supply of a source operand is dependent on a currently
+   * executing instruction. */
+  bool dependent_ = false;
+
+  /** A vector of pairs representing a register dependencies. The first entry in
+   * the pair is the register depended on, and the second is the operand index,
+   * within the dependent instruction, the dependency is associated with. */
+  std::vector<std::pair<Register, uint16_t>> dependency_;
```
[the remark this answers is not in the document]
Comment (jj16791) [reply src/include/simeng/pipeline/BlockingIssueUnit.hh]: Good idea

**Review by FinnWilkinson — CHANGES_REQUESTED** [review]
Final few points, both very minor

File: configs/DEMO_RISCV.yaml [hunk 16]
```diff
@@ -7,6 +7,7 @@ Core:
   ISA: rv64
   Simulation-Mode: outoforder
   Clock-Frequency: 2.5
+  Timer-Frequency: 200
```
[the remark this answers is not in the document]
Comment (jj16791) [reply configs/DEMO_RISCV.yaml:10]: The field has been set as non-optional currently. We can make it optional again but assumed this was now considered to be non-optional.

File: test/unit/pipeline/InOrderStagerTest.cc [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+TEST_F(PipelineInOrderStagerTest, RecordIssue) {
+  inOrderStagerUnit.recordIssue(uopPtr);
+  inOrderStagerUnit.recordIssue(uopPtr2);
+
+  EXPECT_EQ(inOrderStagerUnit.getNextSeqID(), 1);
+  EXPECT_EQ(inOrderStagerUnit.canWriteback(uop->getSequenceId()), true);
+  EXPECT_EQ(inOrderStagerUnit.canWriteback(uop2->getSequenceId()), false);
+  inOrderStagerUnit.recordRetired(uop->getSequenceId());
+  EXPECT_EQ(inOrderStagerUnit.getNextSeqID(), 2);
+  EXPECT_EQ(inOrderStagerUnit.canWriteback(uop2->getSequenceId()), true);
```
[the remark this answers is not in the document]
Comment (jj16791) [reply test/unit/pipeline/InOrderStagerTest.cc:65]: Not sure I understand your point here. The sequence and instruction IDs are set in the constructor of this class so we don't need to set them in each test case?

File: configs/DEMO_RISCV.yaml (hunk 16 above)
[the remark this answers is not in the document]
Comment (FinnWilkinson) [reply configs/DEMO_RISCV.yaml:10]: Its menail anyway. Lets leave it as is

File: test/unit/pipeline/InOrderStagerTest.cc (hunk 17 above)
[the remark this answers is not in the document]
Comment (FinnWilkinson) [reply test/unit/pipeline/InOrderStagerTest.cc:65]: Sorry, didn't see this in the constructor...

File: test/unit/pipeline/InOrderStagerTest.cc [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+TEST_F(PipelineInOrderStagerTest, Flush) {
+  inOrderStagerUnit.recordIssue(uopPtr);
+  inOrderStagerUnit.recordIssue(uopPtr2);
+  inOrderStagerUnit.flush();
+  EXPECT_EQ(inOrderStagerUnit.getNextSeqID(), -1);
+
+  inOrderStagerUnit.recordIssue(uopPtr);
+  inOrderStagerUnit.recordIssue(uopPtr2);
+  inOrderStagerUnit.flush(uop->getInstructionId());
+  EXPECT_EQ(inOrderStagerUnit.getNextSeqID(), uop->getSequenceId());
```
[the remark this answers is not in the document]
Comment (FinnWilkinson) [reply test/unit/pipeline/InOrderStagerTest.cc:78]: Yes you are correct, I was thinking of it the wrong way around

**Review by FinnWilkinson — APPROVED** [review]
All looks good. Thanks for doing this fix / large overhaul!

File: src/include/simeng/models/inorder/Core.hh [hunk 19]
```diff
@@ -140,28 +139,50 @@ class Core : public simeng::Core {
...
+  /** The inorder stager unit; tracks in program-order instructions issued to
+   * ensure writeback functionality occurs in program-order. */
+  pipeline::InOrderStager staging_;
+
+  /** The issue unit; reads operands, and issues ready instructions to the
+   * execution unit in program-order. */
+  pipeline::BlockingIssueUnit issueUnit_;
+
+  /** The set of execution units; executes uops and sends to writeback, also
+   * forwarding results to issue. */
```
Comment (rahahahat) [inline src/include/simeng/models/inorder/Core.hh:170]: Would it be better if the comment said: "...also forwarding results to the issue unit or blocking issue unit"

File: src/lib/pipeline/BlockingIssueUnit.cc [hunk 20]
```diff
@@ -0,0 +1,287 @@
...
+      portAllocator_.deallocate(port);
+      break;
+    }
+
+    // Set scoreboard for all destination registers as not ready
+    for (const auto& reg : destinationRegisters) {
+      scoreboard_[reg.type][reg.tag] = {false, uop->getInstructionId()};
+    }
+
+    // Record that this uop has been issue and send to an execution unit through
```
Comment (rahahahat) [inline src/lib/pipeline/BlockingIssueUnit.cc:123]: "has been issued"

**Review by rahahahat — CHANGES_REQUESTED** [review]

**Comment by rahahahat** [comment]
Very minor changes. All looks good

File: src/include/simeng/models/inorder/Core.hh (hunk 19 above)
Comment (jj16791) [reply src/include/simeng/models/inorder/Core.hh:170]: The comment doesn't gain much from the extra specifics suggested. It's just less concise. Results forwarding is a well-known/defined process so the comment's conciseness works here.

File: src/include/simeng/pipeline/LoadStoreQueue.hh [hunk 21]
```diff
@@ -16,6 +16,9 @@ namespace pipeline {
 /** The memory access types which are processed. */
 enum accessType { LOAD = 0, STORE };
 
+/** The order in which instructions can exit this unit. */
+enum completionOrder { INORDER = 0, OUTOFORDER };
```
[the remark this answers is not in the document]
Comment (jj16791) [reply src/include/simeng/pipeline/LoadStoreQueue.hh:20]: No, can be moved to an `enum class`.

[the remark this answers is not in the document]
Comment (jj16791) [reply src/include/simeng/pipeline/LoadStoreQueue.hh:20]: As with other PR where the same question was asked, no set precedence here for capitalization but I will make the change.

File: src/include/simeng/pipeline/LoadStoreQueue.hh [hunk 22]
```diff
@@ -16,6 +16,9 @@ namespace pipeline {
 /** The memory access types which are processed. */
 enum accessType { LOAD = 0, STORE };
```
[the remark this answers is not in the document]
Comment (jj16791) [reply src/include/simeng/pipeline/LoadStoreQueue.hh:17]: It will need to be cast as its only use is as an integral type (indexing an array.

**Review by rahahahat — APPROVED** [review]

**Comment by jj16791** [comment]
PR is too out of date with `dev`. Closing and will re-open in the future with an updated base
